internal/cluster/distribution/eks/eksworkflow/activity_update_addon.go

```diff
@@ -143,33 +146,63 @@ func (a *UpdateAddonActivity) Execute(ctx context.Context, input UpdateAddonActi
 		}
 		return nil, errors.WrapIfWithDetails(err, "failed to update addon", "cluster", input.ClusterName, "addon", input.AddonName)
 	}
-	output := UpdateAddonActivityOutput{UpdateID: aws.StringValue(updateAddonOutput.Update.Id)}
+	output := UpdateAddonActivityOutput{UpdateID: aws.StringValue(updateAddonOutput.Update.Id), LatestVersion: latestVersion}
 
 	return &output, nil
 }
 
-func selectLatestVersion(addonVersions *eks.DescribeAddonVersionsOutput, currentVersion string, kubernetesVersion string) (string, error) {
+// return latest available version which might be the currentVersion. In case selectNextMinor = true tries to return next available minor version.
+// Beside the next version it returns whether the returned version is the latest one.
+func selectNextVersion(addonVersions *eks.DescribeAddonVersionsOutput, currentVersion string, kubernetesVersion string, selectNextMinor bool) (string, bool, error) {
 	currentVersionSemver, err := semver.NewVersion(currentVersion)
 	if err != nil {
-		return "", err
+		return "", false, err
 	}
 	latestVersion := currentVersionSemver
 
+	versions := make([]*semver.Version, 0, 5)
 	for _, addon := range addonVersions.Addons {
 		for _, version := range addon.AddonVersions {
 			if !versionIsCompatible(version.Compatibilities, kubernetesVersion) {
 				continue
 			}
-			newVersion, err := semver.NewVersion(*version.AddonVersion)
+			version, err := semver.NewVersion(*version.AddonVersion)
 			if err != nil {
-				return "", err
+				return "", false, err
 			}
-			if newVersion.GreaterThan(latestVersion) {
-				latestVersion = newVersion
+			if version.GreaterThan(latestVersion) {
+				versions = append(versions, version)
 			}
 		}
 	}
-	return latestVersion.Original(), nil
+
+	if len(versions) == 0 {
+		return latestVersion.Original(), true, nil
+	}
+
+	sort.Slice(versions, func(i, j int) bool {
+		if versions[i].Compare(versions[j]) == -1 {
+			return true
+		}
+		return false
+	})
+
+	if selectNextMinor {
+		for i, version := range versions {
+			if !isSameMinorVersion(version, latestVersion) {
+				return version.Original(), i+1 == len(versions), nil
+			}
+		}
+	}
+
+	return versions[len(versions)-1].Original(), true, nil
+}
+
+func isSameMinorVersion(newVersion *semver.Version, currentVersion *semver.Version) bool {
+	if newVersion.Major() == currentVersion.Major() && newVersion.Minor() == currentVersion.Minor() {
+		return true
+	}
+	return false
```

Comment:
**Recommended**: this function could be simplified into the following to make it more clean/decrease the cognitive complexity.
```suggestion
	return newVersion.Major() == currentVersion.Major() && 
	    newVersion.Minor() == currentVersion.Minor()
```

**Note**: also probably this condition is not worth the trouble to be put into a function (unless building a semver library), because it is easier to understand the direct 2 part condition than to bear the cognitive load of `isSameMinorVerison() != (x.Minor() == y.Minor())`.

**Note**: with an earlier suggestion implemented this function can even be removed.

Reply:
Fixed in https://github.com/banzaicloud/pipeline/pull/3563/commits/22da7696190597e681af96bcbc1fc21ee85163dc.